demo/odaslive/parameters.c

```diff
@@ -244,6 +288,89 @@
 
     }
 
+    pa_channel_map* parameters_pa_channel_map_config(const char* fileConfig)
+    {
+        bool channelNamesExists = parameters_exists(fileConfig, "raw.interface.channelnames");
+        bool channelMapExists = parameters_exists(fileConfig, "raw.interface.channelmap");
+        if (channelNamesExists && channelMapExists)
+        {
+            printf("Configuration error: raw.interface.channelnames and raw.interface.channelmap cannot be used at the same time\n");
+            exit(EXIT_FAILURE);
+        }
+        else if (!(channelNamesExists || channelMapExists))
+        {
+            printf("Configuration error: either raw.interface.channelnames or raw.interface.channelmap must be used with pulseaudio\n");
+            exit(EXIT_FAILURE);
+        }
+
+        if (channelNamesExists)
+        {
+            char * tmpStr;
+            pa_channel_map* map = (pa_channel_map*) malloc(sizeof(pa_channel_map));
+            pa_channel_map_init(map);
+            map->channels = parameters_lookup_int(fileConfig, "raw.nChannels");
+
+            tmpStr = parameters_lookup_string(fileConfig, "raw.interface.channelnames");
+
+            if (!pa_channel_map_parse(map, tmpStr))
+            {
+                printf("Invalid channel map string: %s\n", tmpStr);
+                exit(EXIT_FAILURE);
+            }
+
+            if (map->channels != parameters_lookup_int(fileConfig, "raw.nChannels"))
+            {
+                printf("Configuration error: raw.interface.channelnames length does not match raw.nChannels\n");
+                exit(EXIT_FAILURE);
+            }
+
+            if (!pa_channel_map_valid(map))
+            {
+                printf("Invalid channel map\n");
+                exit(EXIT_FAILURE);
+            }
+
+            free((void *) tmpStr);
+
+            return map;
+
+        }
+        else
+        {
+            unsigned int nChannels = parameters_count(fileConfig, "raw.interface.channelmap");
+            char tmpStr[1024];
+            pa_channel_map* map = (pa_channel_map*) malloc(sizeof(pa_channel_map));
+            pa_channel_map_init(map);
+            map->channels = 0;
+
+            for (unsigned int iChannel = 0; iChannel < nChannels; ++iChannel)
+            {
+                sprintf(tmpStr, "raw.interface.channelmap.[%u]", iChannel);
+                char* channel = parameters_lookup_string(fileConfig, tmpStr);
+                if (((map->map[iChannel] = pa_channel_position_from_string(channel)) == PA_CHANNEL_POSITION_INVALID))
+                {
+                    printf("Invalid channel position: %s\n", channel);
+                    exit(EXIT_FAILURE);
+                }
+                ++(map->channels);
+                free(channel);
```

Comment:
I don't know why there is "(void *)" for other free usages, but I would add that here to be consistent.

Reply:
Fixed